zeppelin-server/src/main/java/org/apache/zeppelin/rest/LoginRestApi.java

```diff
@@ -100,43 +100,42 @@ public Response getLogin(@Context HttpHeaders headers) {
       if (response == null) {
         Map<String, String> data = new HashMap<>();
         data.put("redirectURL", constructKnoxUrl(knoxJwtRealm, knoxJwtRealm.getLogin()));
-        response = new JsonResponse(Status.OK, "", data);
+        response = new JsonResponse<>(Status.OK, "", data);
       }
       return response.build();
-    } else {
-      KerberosRealm kerberosRealm = getKerberosRealm();
-      if (null != kerberosRealm) {
-        try {
-          Map<String, Cookie> cookies = headers.getCookies();
-          KerberosToken kerberosToken = KerberosRealm.getKerberosTokenFromCookies(cookies);
-          if (null != kerberosToken) {
-            Subject currentUser = org.apache.shiro.SecurityUtils.getSubject();
-            String name = (String) kerberosToken.getPrincipal();
-            if (!currentUser.isAuthenticated() || !currentUser.getPrincipal().equals(name)) {
-              response = proceedToLogin(currentUser, kerberosToken);
-            }
-          }
-          if (null == response) {
-            LOG.warn("No Kerberos token received");
-            response = new JsonResponse(Status.UNAUTHORIZED, "", null);
+    }
+
+    KerberosRealm kerberosRealm = getKerberosRealm();
+    if (null != kerberosRealm) {
+      try {
+        Map<String, Cookie> cookies = headers.getCookies();
+        KerberosToken kerberosToken = KerberosRealm.getKerberosTokenFromCookies(cookies);
+        if (null != kerberosToken) {
+          Subject currentUser = SecurityUtils.getSubject();
+          String name = (String) kerberosToken.getPrincipal();
+          if (!currentUser.isAuthenticated() || !currentUser.getPrincipal().equals(name)) {
+            response = proceedToLogin(currentUser, kerberosToken);
           }
-          return response.build();
-        } catch (AuthenticationException e){
-          LOG.error("Error in Login: " + e);
         }
+        if (null == response) {
+          LOG.warn("No Kerberos token received");
+          response = new JsonResponse<>(Status.UNAUTHORIZED, "", null);
+        }
+        return response.build();
+      } catch (AuthenticationException e){
+        LOG.error("Error in Login", e);
       }
     }
-    return new JsonResponse(Status.METHOD_NOT_ALLOWED).build();
+    return new JsonResponse<>(Status.METHOD_NOT_ALLOWED).build();
   }
 
   private KerberosRealm getKerberosRealm() {
-    Collection realmsList = authenticationService.getRealmsList();
+    Collection<Realm> realmsList = (Collection<Realm>) authenticationService.getRealmsList();
```

Comment:
It's better to change the signature of AuthenticationService instead of using type cast.

Reply:
I'd be happy to do that but hesitated because that's a public API and I didn't know Zeppelins guarantees about changing those.

Reply:
I think it is fine, just go ahead.